### PR TITLE
Added support for Avro schema projection in RecordFormat.

### PR DESCRIPTION
**Pull request by chtyim**

### CODE REVIEW

File: cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java [hunk 1]
```diff
@@ -265,7 +265,8 @@
      * Contains HTTP headers used by Stream handler.
      */
     public static final class Headers {
-      public static final String CONSUMER_ID = "X-ConsumerId";
+      public static final String SCHEMA = "schema";
+      public static final String SCHEMA_HASH = "schema.has";
```
Comment (albertshau) [inline cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java]: typo, has -> hash

Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java]: haha.. good catch.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/format/AvroRecordFormat.java [hunk 2]
```diff
@@ -16,36 +16,51 @@
...
   public GenericRecord read(StreamEvent event) {
     try {
-      return datumReader.read(null, decoderFactory.binaryDecoder(new ByteBufferInputStream(event.getBody()),
-                                                                 binaryDecoder));
+      // Check if the event has different schema then the read schema. If it does update the datumReader
+      String eventSchemaStr = event.getHeaders().get(Constants.Stream.Headers.SCHEMA);
+      String eventSchemaHash = event.getHeaders().get(Constants.Stream.Headers.SCHEMA_HASH);
+      if (eventSchemaStr != null && !this.writeSchemaHash.equals(eventSchemaHash)) {
+        org.apache.avro.Schema writeSchema = new org.apache.avro.Schema.Parser().parse(eventSchemaStr);
+        datumReader.setSchema(writeSchema);
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/format/AvroRecordFormat.java]: the assumption here is that once we see a schema in the headers, every event has that schema until the next event with schema in the headers.  We should document this in the javadocs. It also may make sense to have a setting that disables this behavior if it is not desired.

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/format/AvroRecordFormat.java]: Alternatively we can switch back to use the read schema as the default one if the event doesn't have schema, which probably match better with existing logic.

**Comment by albertshau** [comment]
1 typo and 1 doc comment, otherwise lgtm

**Comment by albertshau** [comment]
:+1:

**Pull request merged**